File: src/main/java/rag/trace/TraceBus.java
package main.java.rag.trace;

import java.util.ArrayList;
import java.util.List;

public class TraceBus {
    private final List<TraceSink> sinks;
    
    public TraceBus() {
        this.sinks = new ArrayList<>();
    }
    
    
    public void register(TraceSink sink) {
        sinks.add(sink);
    }
    
    
    public void unregister(TraceSink sink) {
        sinks.remove(sink);
    }
    

    public void publish(TraceEvent event) {
        for (TraceSink sink : sinks) {
            sink.record(event);
        }
    }
}

File: iteration1_py/src/trace/trace_bus.py
from typing import List
from src.trace.trace_sink_abstract import TraceSinkAbstract
from src.trace.trace_event import TraceEvent


class TraceBus:
    
    def __init__(self):
        self.__sinks: List[TraceSinkAbstract] = []
    
    def register(self, sink: TraceSinkAbstract) -> None:
        self.__sinks.append(sink)
    
    def unregister(self, sink: TraceSinkAbstract) -> None:
        if sink in self.__sinks:
            self.__sinks.remove(sink)
    
    def publish(self, event: TraceEvent) -> None:
        for sink in self.__sinks:
            sink.record(event)
